src/python/pants/backend/python/goals/repl.py

```diff
@@ -46,6 +68,7 @@ async def create_python_repl_request(repl: PythonRepl, pex_env: PexEnvironment)
     extra_env = {
         **complete_pex_env.environment_dict(python_configured=requirements_pex.python is not None),
         "PEX_EXTRA_SYS_PATH": ":".join(chrooted_source_roots),
+        "PEX_PATH": repl.in_chroot(local_dists.pex.name),
```

Comment:
So now you depend on the `local_dists` twice. Once up in `PEX_EXTRA_SYS_PATH` above as loose sources (with no `.dist-info/` metadata dir) and then here as an installed wheel with a `.dist-info/` metadata dir. I'm not sure which is 1st on the final `sys.path` and the order should not matter anyhow since there will be just one `proj-rev.dist-info/` dir on the `PYTHONPATH`, but ... this duplication of everything else feels like we're walking into a trap. For example, although the `.dist-info/` case works out solving one user's issue, I'm pretty sure a native extension that lives in a package may not. For example, if there is pure python code packaged as `package/module.py` and `package/module.py` does not live in a maespace package and the loose source version of this file ends up 1st on the `sys.path`, then the `package/native.cpython-39-x86_64-linux-gnu.so` located in an installed wheel on the `PEX_PATH` will never be found.

Reply:
Yeah:
```console
$ cat <<EOF > BUILD.test
python_library(
  name="jake",
  dependencies=[
    "src/python/pants:pants-packaged",
  ],
)
EOF
^jsirois@gill ~/dev/pantsbuild/pants (main) $ ./pants filedeps --transitive //:jake
3rdparty/python/BUILD
3rdparty/python/requirements.txt
BUILD.test
src/python/pants/BUILD
src/python/pants/VERSION
src/python/pants/__main__.py
src/python/pants/backend/awslambda/python/BUILD
src/python/pants/backend/awslambda/python/__init__.py
src/python/pants/backend/awslambda/python/lambdex.py
src/python/pants/backend/awslambda/python/register.py
...
```

Reply:
Yes, that is correct, and in fact occurred during my testing. The loose sources win, and make the native code unimportable if not in a namespace package. 

The duplication itself can be mitigated using the `!!` operator to exclude deps. And we can mitigate it in the future by  A) not attempting to infer deps that can be satisfied by a local dist and B) not following explicit deps through a `python_distribution` target when computing the transitive closure. 

But the namespace package problem may occur regardless, even if we eliminate duplication, because of code outside the dist sharing a package with code in the dist. So depending on a dist like this does require you to know what you're doing.

That said, as this stands, it solves a real user problem, albeit with a gotcha, which we can document. So, on balance, we might want to release this and potentially deal with the missing pieces as above in a followup. WDYT?

Reply:
I have no well justified thoughts. It seems to me subtracting the closure of a python_distribution dep sub graph from the larger graph in play is not actually hard, in which case it would be great to just take a day or a week to do things right and ship in the next dev release. I may be underestimating the difficulty of doing this right from the get go though.

Reply:
Yeah, I'm not sure how hard it would be. But it doesn't fully solve the namespace problem. In most/many cases you will still have to ensure the code is in a namespace package, at which point the duplication problem seems less critical.

Reply:
Ok. I'm far less confident this doesn't lead to other problems, but sgtm.

Reply:
Yeah, you are quite probably right... But I guess this is mitigated by the fact that this only impacts users who manually create a dep on a `python_distribution`, not users in general. We can treat this as an advanced feature where there be dragons, and see how that goes? Also, will open this up to other opinions.

Reply:
@stuhood @illicitonion do you have any thoughts? Not so much on the code, as on the potential impact, re the discussion above.

Reply:
This goes back to the recent slack thread about why Bazel has both `deps` and `data`: a consumer classifies how the dep should be treated... in this case, the `pytest` rules could have a special argument for these dependencies to indicate that they should be treated as installed distributions rather than sources.

> not following explicit deps through a python_distribution target when computing the transitive closure.

Pants has started to go down this route via `special_cased_dependencies`, although the support is nascent (...as the name suggests), and is not type-aware. See https://github.com/pantsbuild/pants/pull/12251#discussion_r667079976 for some discussion and see [the implementation](https://github.com/pantsbuild/pants/blob/7a5205aa23874bd67f7b00acd6b060684d778e0d/src/python/pants/engine/internals/graph.py#L866-L871).

That parameter essentially amounts to a boolean filter of dependencies that is applied while computing either direct or transitive dependencies. I say that it's "not type aware" because only a consumer who is aware that a particular target type has `SpecialCasedDependencies`-subclassed fields will actually be able to consume them "correctly", whereas I could imagine an API where `DependenciesRequest` or `TransitiveTargetRequest` optionally take a list of fields to expand dependencies from (with a default of just `Dependencies`), and the more widely used they are, the more likely that is to be necessary.

@Eric-Arellano would know more. But it seems like two options would be:
1. `python_distribution` marking its own `dependencies` field `SpecialCasedDependencies` and adjusting how it requests them
    * this would impact all consumers of `python_distribution` targets: they would not (by default) get the full transitive deps of that target: they'd instead get "whatever behavior we want when a `python_library` depends on a `python_distribution`" 
2. `python_tests`/`python_library`/etc gaining a special-cased field to which you could add things that you wanted to treat as installed distributions rather than sources.

I think that given that we have the `python_distribution` target which already "marks" a particular intended behavior, option 2 would be redundant. Option 1 would also be more consistent (haven't looked at the code, but I think it's the path that you're already on here): support for Python code consuming `python_distribution`s would need to be added ~everywhere.

Reply:
Yeah, I continue to think it's important to distinguish between "data" vs "deps". 

Thus far, everything defined in the `dependencies` field behaves fairly simply and consistently: we copy all relevant source files (possibly generated), and also add deps to all transitive deps. In the case of `python_distribution`, it does not have `sources`, so we simply add all its transitive deps.

I'm concerned about now having to keep track of all the different unwritten semantics. For example, this would be confusing imo:

- if you depend on most targets, you get the source code + deps
- if you depend on `pex_binary`, we'll build the PEX and insert at `<path>`. We won't include transitive deps.
- if you depend on `python_distribution`, we'll not only build it but also install it. We won't include transitive deps.

I fear it's often not apparent just by looking at target names in a BUILD file what is what:

```python
# What target is what? You have to either 
# use `./pants peek` or jump to BUILD files to find out.
python_tests(
   dependencies=[
      "src/py/project",
      "src/py/project:app",
      "src/py/project:lib",
],
```

Sure, people often use `pex-binary`, `lambda`, and `dist` to name things, but we don't enforce that nor even recommend it.

--

I acknowledge having a "special cased" field could be seen as clunky, but I think there's value in its explicitness:

- it's more clear when reading a BUILD file what will have magical handling. Strawman naming:

```python
python_tests(
   dependencies=["src/py/project:lib"],
   local_dists=["src/py/project"],
   runtime_package_dependencies=["src/py/project:app"],
)
```
- `./pants help` can explain how the special logic behaves—e.g. where `pex_binary` gets installed to. 

--

Special-cased fields also gives us much more flexibility: you can now use the same target type for multiple consumption purposes:

> what can a test do with a .whl file?

In your test, you can unzip the .whl and inspect that everything is there that you expect. In fact, Liam was working on a test to do this for PEP 561's `py.typed` file.

By having special-cased fields, you could use the same `python_distribution` to both a) install as a local dist you can import, b) inspect the `.whl` file, and c) use in some way we can't anticipate that plugin authors write. Or, pick and choose from those 3.

Reply:
I've proposed this elsewhere and it's a handwave, but I think you can fill in the blanks. Same thrust as Eric, but probably more flexible, less boilerplate:
```python
python_tests(
    dependencies=[
        "src/py/project:lib",
        "src/py/project@installed_dist",
        "src/py/project:app@binary",
    ],
)
```

Here rules can register things like `installed_dist` and `binary` and then we have wiring that ensures those rules get invoked and their output files get added to consuming chroots.

Reply:
Thanks for the thoughtful comments! 

I suspect that the behavior Eric describes as "simple and consistent" only seems so to us, because we are so used to thinking in these terms.  The behavior Eric describes as "confusing" is, I think, exactly what users both expect and want. 

For example, I think that to end users it would be pretty intuitive that depending on a `python_distribution` means building it and installing it into the environment of the dependee. 

This gets to the fact that we have - in practice - "inner" or "code" targets that own files (python_library, python_tests, files, resources), and then "artifact" targets (python_distribution, pex_binary) that don't own files but instead represent an artifact to be built. We don't model this distinction explicitly, but we probably should. Maybe we should even use different terms for them.  Depending on an "artifact target" can mean something different than depending on a "code target", and I think that meaning is fairly unambiguous: depending on an edge target means depending on whatever is built from it, and *not* depending transitively on its own deps. 

I think this is a more user-centric way of thinking about it than requiring the burden of a special-case field. 

> In your test, you can unzip the .whl and inspect that everything is there that you expect. 

That is not useful, unless you're testing Pants itself. A built pex_binary may be "data", in some sense, but a built python_distribution is not - it has to be installed into the environment.  Concretely, if you have a native extension you expect to be able to use it in your tests. That requires building and installing the dist. Making people jump through weird hoops here seems not great.

The complicating issue, no matter how we end up supporting this, is what to do with dep inference. If a symbol is provided by the built artifact, that we don't want to also infer a dep on the source of that symbol. And this will be the case even if we use some special-cased field.

Reply:
>  "inner" or "code" targets that own files (python_library, python_tests, files, resources), and then "artifact" targets (python_distribution, pex_binary) that don't own files but instead represent an artifact to be built. We don't model this distinction explicitly, but we probably should. Maybe we should even use different terms for them.

This is precisely why I have been so excited about that cleanup a few months ago for `python_awslambda` and `pex_binary` to no longer have `sources`. This is the mental model I've been using since that breakthrough, including when envisioning the elusive "re-envisioning targets" proposal. I do think there's a big difference between those two types of targets, and it is valuable to codify it more.

Still handwavy, but imo: targets work really well for artifact-type targets like `pex_binary`. They also seem to work well for 3rd-party deps metadata. They do not work well for first-party code, with issues like it being too hard to update just 1 file's metadata. So, I could envision us codifying this distinction by replacing 1st-party targets w/ something else entirely, for example.

--

>  about now having to keep track of all the different unwritten semantics.

@benjyw it sounds like you are envisioning a particular target always behaving the same way when in `dependencies`, and not depending on the call site's custom semantics? (Outside of project introspection). So, if you depend on a `python_distribution`, that _always_ means to install it? 

That consistency would be good imo and make me more comfortable with this.

If so, where would that be documented? On the `python_distribution` target itself? In the `dependencies` field `help` message for each consumer? (I'm asking because I think it's important we know how to explain this to users.)

Reply:
> By having special-cased fields, you could use the same `python_distribution` to both a) install as a local dist you can import, b) inspect the `.whl` file, and c) use in some way we can't anticipate that plugin authors write. Or, pick and choose from those 3.

To clarify something: _both_ of my suggestions involved using the `SpecialCasedDependencies` field, but they proposed using them in different places. Would suggest taking another look there.

The current implementation allows for _either_ "consumer decides" _or_ "producer decides", depending on where `SpecialCasedDependencies` are used.

> @benjyw it sounds like you are envisioning a particular target always behaving the same way when in dependencies, and not depending on the call site's custom semantics? (Outside of project introspection). So, if you depend on a python_distribution, that always means to install it?

This is also what I felt would likely be least redundant:
> I think that given that we have the python_distribution target which already "marks" a particular intended behavior, option 2 would be redundant.

Reply:
Yes - I agree that we always want the same behavior when you depend on a `python_distribution`, and I think that behavior fairly obviously should be "build the dist and install it in the downstream environment".  Having a dependency mean "depend on the python_library targets that a python_distribution depends on", seems weird. And if someone really wants to depend on those python_library targets, they can just depend on them directly. 

In basically 100% of cases today, depending on a `python_distribution` is a smell, and we have masked the smell by giving that dependency a not-particularly-useful meaning. And I doubt there are many such dependencies today, since it doesn't really make sense, especially with dep inference. What I'm proposing is to have that dependency mean something more intuitive.

That said, again, dep inference really muddies the water...

Reply:
> So I think you now depend on the `local_dists` twice. Once up in `PEX_EXTRA_SYS_PATH` above as loose sources (with no `.dist-info/` metadata dir) and then here as an installed wheel with a `.dist-info/` metadata dir.

This still seems like a blocking issue. It seems like you need to introduce `SpecialCasedDependencies` for the distribution target.

Reply:
Yes, but that breaks other things.